http/vibe/http/fileserver.d

```diff
@@ -709,3 +678,55 @@ unittest
 	assert(!cacheMatch(`"xyzzy",W/"r2d2xxxx", "c3piozzzz"`, ETag(No.weak, "r2d2xxxx"), No.allowWeak));
 	assert( cacheMatch(`"xyzzy",W/"r2d2xxxx", "c3piozzzz"`, ETag(No.weak, "r2d2xxxx"), Yes.allowWeak));
 }
+
+private RangeSpec parseRangeHeader(string range_spec, ulong file_size, scope HTTPServerResponse res)
+{
+	RangeSpec ret;
+
+	auto range = range_spec.chompPrefix("bytes=");
+	if (range.canFind(','))
+		throw new HTTPStatusException(HTTPStatus.notImplemented);
+	auto s = range.split("-");
+	if (s.length != 2)
+		throw new HTTPStatusException(HTTPStatus.badRequest);
+
+	// https://tools.ietf.org/html/rfc7233
+	// Range can be in form "-\d", "\d-" or "\d-\d"
+	try {
+		if (s[0].length) {
+			ret.min = s[0].to!ulong;
+			ret.max = s[1].length ? s[1].to!ulong + 1 : file_size;
+		} else if (s[1].length) {
+			ret.min = file_size - min(s[1].to!ulong, file_size);
+			ret.max = file_size;
+		} else {
+			throw new HTTPStatusException(HTTPStatus.badRequest);
+		}
+	} catch (ConvException) {
+		throw new HTTPStatusException(HTTPStatus.badRequest);
+	}
+
+	if (ret.max > file_size) ret.max = file_size;
+
+	if (ret.min >= ret.max) {
+		res.headers["Content-Range"] = "bytes */%s".format(file_size);
+		throw new HTTPStatusException(HTTPStatus.rangeNotSatisfiable);
+	}
+
+	return ret;
+}
+
+unittest {
+	auto res = createTestHTTPServerResponse();
+	assertThrown(parseRangeHeader("bytes=2-1", 10, res));
+	assertThrown(parseRangeHeader("bytes=10-10", 10, res));
```

Comment:
I believe this is a valid range (10-10 would simply mean, just the one byte at index `10`) and thus should not throw.

See 2.1 in the RFC which only says:

```
   A byte-range-spec is invalid if the last-byte-pos value is present
   and less than the first-byte-pos.
```

plus it's actually used as an example here:

```
   o  The first and last bytes only (bytes 0 and 9999):
        bytes=0-0,-1
```

Reply:
Wait, but index 10 is outside of 10 bytes isn't it?

Reply:
Or do you mean that it should get moved into the range as `9-9`?

Reply:
oh sorry you are right about that, I wasn't looking at the size there